Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,312 +41,588 @@ public sealed class RCDSystem : EntitySystem
+        {
+            if (popMsgs)
+                _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-on-non-empty-tile-message"), uid, user);
+
             return false;
+        }
 
-        var unobstructed = target == null
-            ? _interaction.InRangeUnobstructed(user, mapGrid.GridTileToWorld(tile.GridIndices), popup: true)
-            : _interaction.InRangeUnobstructed(user, target.Value, popup: true);
+        // Check rule: Must build on non-empty tile
+        if (!component.CachedPrototype.ConstructionRules.Contains(RcdConstructionRule.CanBuildOnEmptyTile) && mapGridData.Tile.Tile.IsEmpty)
+        {
+            if (popMsgs)
+                _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-on-empty-tile-message"), uid, user);
 
-        if (!unobstructed)
             return false;
+        }
 
-        switch (comp.Mode)
+        // Check rule: Must place on subfloor
+        if (component.CachedPrototype.ConstructionRules.Contains(RcdConstructionRule.MustBuildOnSubfloor) && !mapGridData.Tile.Tile.GetContentTileDefinition().IsSubFloor)
         {
-            //Floor mode just needs the tile to be a space tile (subFloor)
-            case RcdMode.Floors:
-                if (!tile.Tile.IsEmpty)
+            if (popMsgs)
+                _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-as-subfloor-not-exposed-message"), uid, user);
+
+            return false;
+        }
+
+        // Tile specific rules
+        if (component.CachedPrototype.Mode == RcdMode.ConstructTile)
+        {
+            // Check rule: Tile placement is valid
+            if (!_floors.CanPlaceTile(mapGridData.GridUid, mapGridData.Component, out var reason))
+            {
+                if (popMsgs)
+                    _popup.PopupClient(reason, uid, user);
+
+                return false;
+            }
+
+            // Check rule: Tiles can't be identical
+            if (mapGridData.Tile.Tile.GetContentTileDefinition().ID == component.CachedPrototype.Prototype)
+            {
+                if (popMsgs)
+                    _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-as-tiles-are-identical"), uid, user);
+
+                return false;
+            }
+
+            // Ensure that all construction rules shared between tiles and object are checked before exiting here
+            return true;
+        }
+
+        // Object specific rules
+
+        // Check rule: The tile is unoccupied
+        var isWindow = component.CachedPrototype.ConstructionRules.Contains(RcdConstructionRule.IsWindow);
+        var constructionAngle = (component.CachedPrototype.Rotation == RcdRotation.User) ? component.ConstructionDirection.ToAngle() : Direction.South.ToAngle();
+
+        foreach (var ent in _lookup.GetLocalEntitiesIntersecting(mapGridData.Tile, -0.05f, LookupFlags.Uncontained))
+        {
```

Comment:
Use the overload that lets you pass in a hashset and store it as a field on the system to avoid having to re-allocate it every time.